File: CSharp/Legacy89DiskKit.Infrastructure/Fdc/Medium/SectorBackedControllerFacingMedium.cs
using Legacy89DiskKit.Domain.Fdc.Interface;

namespace Legacy89DiskKit.Infrastructure.Fdc.Medium;

public abstract class SectorBackedControllerFacingMedium : IControllerFacingMedium
{
    private byte _status;
    private byte _track;
    private byte _sector = 1;
    private byte _data;
    private int _selectedSide;
    private bool _irq;
    private bool _drq;
    private byte[] _transferBuffer = [];
    private int _transferIndex;

    public abstract string MediumKind { get; }

    public bool IsReady => true;

    public abstract bool IsWriteProtected { get; }

    public int SelectedSide => _selectedSide;

    public bool IsIrqAsserted => _irq;

    public bool IsDrqAsserted => _drq;

    public void Reset()
    {
        _status = 0;
        _track = 0;
        _sector = 1;
        _data = 0;
        _selectedSide = 0;
        _irq = false;
        _drq = false;
        ClearTransfer();
    }

    public void SelectSide(int side)
    {
        _selectedSide = side;
    }

    public void SeekTrack(int track)
    {
        _track = (byte)Math.Clamp(track, byte.MinValue, byte.MaxValue);
    }

    public byte ReadStatus()
    {
        return _status;
    }

    public byte ReadTrackRegister()
    {
        return _track;
    }

    public byte ReadSectorRegister()
    {
        return _sector;
    }

    public byte PeekDataRegister()
    {
        return _data;
    }

    public byte ReadDataRegister()
    {
        var value = _data;

        if (_drq)
        {
            AdvanceTransfer();
        }

        return value;
    }

    public void WriteCommand(byte value)
    {
        if ((value & 0xF0) == 0x80)
        {
            ExecuteReadSector();
            return;
        }

        switch (value)
        {
            case <= 0x0F:
                ExecuteRestore();
                break;
            case >= 0x10 and <= 0x1F:
                ExecuteSeek();
                break;
            case 0xD0:
                ExecuteForceInterrupt();
                break;
            default:
                _status = 0x40;
                _irq = true;
                _drq = false;
                ClearTransfer();
                break;
        }
    }

    public void WriteTrackRegister(byte value)
    {
        _track = value;
    }

    public void WriteSectorRegister(byte value)
    {
        _sector = value;
    }

    public void WriteDataRegister(byte value)
    {
        _data = value;
    }

    protected abstract bool SectorExistsCore(int track, int side, int sector);

    protected abstract byte[] ReadSectorCore(int track, int side, int sector);

    private void ExecuteRestore()
    {
        _track = 0;
        _status = 0;
        _irq = true;
        _drq = false;
        ClearTransfer();
    }

    private void ExecuteSeek()
    {
        _track = _data;
        _status = 0;
        _irq = true;
        _drq = false;
        ClearTransfer();
    }

    private void ExecuteForceInterrupt()
    {
        _status = 0;
        _irq = false;
        _drq = false;
        ClearTransfer();
    }

    private void ExecuteReadSector()
    {
        if (!SectorExistsCore(_track, _selectedSide, _sector))
        {
            _status = 0x10;
            _irq = true;
            _drq = false;
            ClearTransfer();
            return;
        }

        var data = ReadSectorCore(_track, _selectedSide, _sector);
        _status = 0;
        _irq = true;
        BeginTransfer(data);
    }

    private void BeginTransfer(byte[] data)
    {
        _transferBuffer = data.Length == 0 ? [0] : data;
        _transferIndex = 0;
        _data = _transferBuffer[0];
        _drq = true;
    }

    private void AdvanceTransfer()
    {
        _transferIndex++;
        if (_transferIndex >= _transferBuffer.Length)
        {
            _drq = false;
            _data = 0;
            ClearTransfer();
            return;
        }

        _data = _transferBuffer[_transferIndex];
        _drq = true;
    }

    private void ClearTransfer()
    {
        _transferBuffer = [];
        _transferIndex = 0;
    }
}

File: Cpp/Legacy89DiskKit.Cpp/src/infrastructure/fdc/medium/sector_backed_controller_facing_medium.cpp
#include "legacy89diskkit/cpp/infrastructure/fdc/medium/sector_backed_controller_facing_medium.hpp"

#include <algorithm>

namespace legacy89diskkit::cpp
{
bool SectorBackedControllerFacingMedium::IsReady() const
{
    return true;
}

int SectorBackedControllerFacingMedium::SelectedSide() const
{
    return selected_side_;
}

bool SectorBackedControllerFacingMedium::IsBusy() const
{
    return pending_operation_ != PendingOperation::None;
}

std::optional<std::chrono::nanoseconds> SectorBackedControllerFacingMedium::GetPendingDelayHint() const
{
    if (pending_operation_ == PendingOperation::None)
    {
        return std::nullopt;
    }

    return remaining_delay_;
}

bool SectorBackedControllerFacingMedium::IsIrqAsserted() const
{
    return irq_;
}

bool SectorBackedControllerFacingMedium::IsDrqAsserted() const
{
    return drq_;
}

void SectorBackedControllerFacingMedium::Reset()
{
    status_ = 0;
    track_ = 0;
    sector_ = 1;
    data_ = 0;
    selected_side_ = 0;
    irq_ = false;
    drq_ = false;
    pending_operation_ = PendingOperation::None;
    remaining_delay_ = std::chrono::nanoseconds::zero();
    pending_seek_track_ = 0;
    ClearTransfer();
}

void SectorBackedControllerFacingMedium::Advance(const std::chrono::nanoseconds delta)
{
    if (pending_operation_ == PendingOperation::None || delta <= std::chrono::nanoseconds::zero())
    {
        return;
    }

    remaining_delay_ -= delta;
    if (remaining_delay_ > std::chrono::nanoseconds::zero())
    {
        return;
    }

    CompletePendingOperation();
}

void SectorBackedControllerFacingMedium::SelectSide(const int side)
{
    selected_side_ = side;
}

void SectorBackedControllerFacingMedium::SeekTrack(const int track)
{
    track_ = static_cast<std::uint8_t>(std::clamp(track, 0, 255));
}

std::uint8_t SectorBackedControllerFacingMedium::ReadStatus()
{
    return status_;
}

std::uint8_t SectorBackedControllerFacingMedium::ReadTrackRegister() const
{
    return track_;
}

std::uint8_t SectorBackedControllerFacingMedium::ReadSectorRegister() const
{
    return sector_;
}

std::uint8_t SectorBackedControllerFacingMedium::PeekDataRegister() const
{
    return data_;
}

std::uint8_t SectorBackedControllerFacingMedium::ReadDataRegister()
{
    const auto value = data_;
    if (drq_)
    {
        AdvanceTransfer();
    }

    return value;
}

void SectorBackedControllerFacingMedium::WriteCommand(const std::uint8_t value)
{
    if ((value & 0xf0) == 0x80)
    {
        StartPendingOperation(PendingOperation::ReadSector);
        return;
    }

    if (value <= 0x0f)
    {
        StartPendingOperation(PendingOperation::Restore);
        return;
    }

    if (value >= 0x10 && value <= 0x1f)
    {
        pending_seek_track_ = data_;
        StartPendingOperation(PendingOperation::Seek);
        return;
    }

    if (value == 0xd0)
    {
        ExecuteForceInterrupt();
        return;
    }

    status_ = FdcStatusUnsupportedCommand;
    irq_ = true;
    drq_ = false;
    ClearTransfer();
}

void SectorBackedControllerFacingMedium::WriteTrackRegister(const std::uint8_t value)
{
    track_ = value;
}

void SectorBackedControllerFacingMedium::WriteSectorRegister(const std::uint8_t value)
{
    sector_ = value;
}

void SectorBackedControllerFacingMedium::WriteDataRegister(const std::uint8_t value)
{
    data_ = value;
}

void SectorBackedControllerFacingMedium::StartPendingOperation(const PendingOperation operation)
{
    pending_operation_ = operation;
    remaining_delay_ = kCommandDelay;
    status_ = FdcStatusBusy;
    irq_ = false;
    drq_ = false;
    ClearTransfer();
}

void SectorBackedControllerFacingMedium::CompletePendingOperation()
{
    const auto operation = pending_operation_;
    pending_operation_ = PendingOperation::None;
    remaining_delay_ = std::chrono::nanoseconds::zero();

    switch (operation)
    {
    case PendingOperation::None:
        return;
    case PendingOperation::Restore:
        track_ = 0;
        status_ = FdcStatusNone;
        irq_ = true;
        drq_ = false;
        ClearTransfer();
        return;
    case PendingOperation::Seek:
        track_ = pending_seek_track_;
        status_ = FdcStatusNone;
        irq_ = true;
        drq_ = false;
        ClearTransfer();
        return;
    case PendingOperation::ReadSector:
        CompleteReadSector();
        return;
    }
}

void SectorBackedControllerFacingMedium::ExecuteForceInterrupt()
{
    status_ = FdcStatusNone;
    irq_ = false;
    drq_ = false;
    pending_operation_ = PendingOperation::None;
    remaining_delay_ = std::chrono::nanoseconds::zero();
    ClearTransfer();
}

void SectorBackedControllerFacingMedium::CompleteReadSector()
{
    if (!SectorExistsCore(track_, selected_side_, sector_))
    {
        status_ = FdcStatusRecordNotFound;
        irq_ = true;
        drq_ = false;
        ClearTransfer();
        return;
    }

    status_ = FdcStatusNone;
    irq_ = true;
    BeginTransfer(ReadSectorCore(track_, selected_side_, sector_));
}

void SectorBackedControllerFacingMedium::BeginTransfer(std::vector<std::uint8_t> data)
{
    transfer_buffer_ = data.empty() ? std::vector<std::uint8_t>{0} : std::move(data);
    transfer_index_ = 0;
    data_ = transfer_buffer_.front();
    drq_ = true;
}

void SectorBackedControllerFacingMedium::AdvanceTransfer()
{
    ++transfer_index_;
    if (transfer_index_ >= transfer_buffer_.size())
    {
        drq_ = false;
        data_ = 0;
        ClearTransfer();
        return;
    }

    data_ = transfer_buffer_[transfer_index_];
    drq_ = true;
}

void SectorBackedControllerFacingMedium::ClearTransfer()
{
    transfer_buffer_.clear();
    transfer_index_ = 0;
}
}
